databuilder/query_language.py

```diff
@@ -31,16 +31,19 @@ class _DictArg(list):
 
 
 class Dataset:
+    def __init__(self):
+        object.__setattr__(self, "variables", {})
+
     def set_population(self, population):
         validate_population_definition(population.qm_node)
-        object.__setattr__(self, "population", population)
+        self.variables["population"] = population
 
     def __setattr__(self, name, value):
         if name == "population":
```

Comment:
I think this means `variables` is now a reserved (albeit silly) name so we should raise an error if it's used. Which makes me think that `_variables` would be a better name, and then we could reject any names starting with an underscore.

Reply:
Oh, that's a good idea -- rules out some other unlikely errors as well.

Reply:
Actually, I tell a lie. When I came to implement this I realized that I did think about this and decided it was better not to use an underscore because it's used outside the class (in `compile()`). Your point about it being a reserved word stands though, so I will check for it.

Reply:
Ah yes, well actually I've been meaning to move `compile()` to `Dataset._compile()`! I don't think we benefit at all from it being a separate function. It would be much nicer if we could say that a dataset is just an object with a `_compile` method that returns query model.

Reply:
I think we separated them originally so that the compile method wasn't available (or not obviously available) to dataset definition writers. Giving it an underscore sort of solves that -- but at the cost of having a different client (`main`) call an underscore method.